### PR TITLE
allow zip options to be set on fs.add*()

### PR DESCRIPTION
**Pull request by mudcube**
Currently there is no way to set a specific zip level while using fs.addBlob, addText, addData64URI. This pull request is a slight tweak that allows a developer to set the zip level, comment, or lastModDate.

Example:

```
fs.addBlob(name, blob, {level: 0});
```

This is useful to me because it's nice to be able to exclude PNGs, OGGs and other already compressed data from being further compressed, while still being able to compress JSON, and package into one zip blob.  btws, thanks for the wonderful package!

### CODE REVIEW

File: WebContent/zip-fs.js [hunk 1]
```diff
@@ -141,10 +141,7 @@
 					}, function(index) {
 						if (onprogress)
 							onprogress(currentIndex + index, totalSize);
-					}, {
-						directory : child.directory,
-						version : child.zipVersion
-					});
```
Comment (Rob--W) [inline WebContent/zip-fs.js]: Could you revert the change to these lines and add a new key:

```js
level: child.compressionLevel
```

File: WebContent/zip-fs.js [hunk 2]
```diff
@@ -301,7 +298,9 @@
 			that.id = fs.entries.length;
 			that.parent = parent;
 			that.children = [];
-			that.zipVersion = params.zipVersion || 0x14;
+			that.options = params.options || {};
+			that.options.version = params.zipVersion || 0x14;
+			that.options.directory = that.directory;
```
Comment (Rob--W) [inline WebContent/zip-fs.js]: And change these lines to

```js
that.compressionLevel = params.compressionLevel;
```

(your current method has an issue: it takes and modifies the user-supplied `options` object)

**Comment by mudcube** [comment]
Hi Rob, 
Thanks for the quick response! 

Do you think it makes sense to also support the 'comment' and 'lastModDate' options as well? I don't personally use them, but that was the idea behind sending the 'options' object in addBlob().

Also, do you think it'd make more sense for the param to be 'level' instead of 'compressionLevel' to match the docs on 'ZipWriter.prototype.add'? Or since it's an abstraction you'd prefer to go with the more precise naming?

You make a good point that modifying the user-supplied options is not good practice and will modify that as requested.

**Comment by Rob--W** [comment]
> Also, do you think it'd make more sense for the param to be 'level' instead of 'compressionLevel' to match the docs on 'ZipWriter.prototype.add'? Or since it's an abstraction you'd prefer to go with the more precise naming?

Stick to `compressionLevel`, because it's more obvious. The parameters don't exactly match anyway, e.g. see `zipVersion` vs `version` in the parameters.

I also typed another comment but it wasn't submitted for some reason: `params` is already a kind of options object, so adding yet another "options" object is a bit weird. I suggest to merge the keys from the function call with the params object, either by explicitly spelling our the supported option keys, or by merging all keys (this is a generic way, see the example below):

```js
    // Around line 282 in zip-fs.js
    function addChild(parent, name, params, extraParams, directory) {
        if (extraParams) {
            for (var key in extraParams) {
                if (Object.prototype.hasOwnProperty.call(extraParams, key)) {
                    params[key] = extraParams[value];
                }
            }
        }
        if (parent.directory)
            return directory ? new ZipDirectoryEntry(parent.fs, name, params, parent) : new ZipFileEntry(parent.fs, name, params, parent);
        else
            throw "Parent entry is not a directory.";
    }
```

Used like:

```js
    ZipDirectoryEntryProto.addBlob = function(name, blob, param) {
        return addChild(this, name, {
            data : blob,
            Reader : BlobReader,
            Writer : BlobWriter
        }, params);
    };
```

After finalizing your patch, please submit another update to the documentation at the `gh-pages` branch, so that your new feature becomes documented.

**Comment by gildas-lormeau** [comment]
closed as obsolete (and fixed)
